libwebauthn/src/webauthn.rs

```diff
@@ -248,60 +249,84 @@ where
-                public_key,
-                ctap2_request.permissions(),
-                ctap2_request.permissions_rpid(),
-            )
+        // For operations that include a PIN, we want to fetch one before obtaining a shared secret.
+        // This prevents the shared secret from expiring whilst we wait for the user to enter a PIN.
+        let pin = match uv_operation {
+            Ctap2UserVerificationOperation::None => unreachable!(),
+            Ctap2UserVerificationOperation::GetPinToken
+            | Ctap2UserVerificationOperation::GetPinUvAuthTokenUsingPinWithPermissions => {
+                Some(obtain_pin(channel, pin_provider, timeout).await?)
+            }
+            Ctap2UserVerificationOperation::GetPinUvAuthTokenUsingUvWithPermissions => {
+                None // TODO probably?
+            }
+        };
+
+        // In preparation for obtaining pinUvAuthToken, the platform:
+        // * Obtains a shared secret.
+        let uv_proto = select_uv_proto(&get_info_response).await?;
+        let (public_key, shared_secret) = obtain_shared_secret(channel, &uv_proto, timeout).await?;
+
+        // Then the platform obtains a pinUvAuthToken from the authenticator, with the mc (and likely also with the ga)
+        // permission (see "pre-flight", mentioned above), using the selected operation.
+        let token_request = match uv_operation {
+            Ctap2UserVerificationOperation::None => unreachable!(),
+            Ctap2UserVerificationOperation::GetPinToken => {
+                Ctap2ClientPinRequest::new_get_pin_token(
+                    uv_proto.version(),
+                    public_key,
+                    &uv_proto.encrypt(&shared_secret, &pin_hash(&pin.unwrap()))?,
+                )
+            }
+            Ctap2UserVerificationOperation::GetPinUvAuthTokenUsingPinWithPermissions => {
+                Ctap2ClientPinRequest::new_get_pin_token_with_perm(
+                    uv_proto.version(),
+                    public_key,
+                    &uv_proto.encrypt(&shared_secret, &pin_hash(&pin.unwrap()))?,
+                    ctap2_request.permissions(),
+                    ctap2_request.permissions_rpid(),
+                )
+            }
+            Ctap2UserVerificationOperation::GetPinUvAuthTokenUsingUvWithPermissions => {
+                Ctap2ClientPinRequest::new_get_uv_token_with_perm(
+                    uv_proto.version(),
+                    public_key,
+                    ctap2_request.permissions(),
+                    ctap2_request.permissions_rpid(),
+                )
+            }
+        };
+
+        match channel.ctap2_client_pin(&token_request, timeout).await {
+            Ok(t) => {
+                break (uv_proto, t, shared_secret);
+            }
+            // Internal retry, because we otherwise can't fall back to PIN, if the UV is blocked
+            Err(Error::Ctap(CtapError::UvBlocked)) => {
+                uv_blocked = true;
```

Comment:
It may be worth adding a `warn!` or `info!` here

Reply:
True